src/site/content/en/mini-apps/mini-app-open-source-projects/index.md

```diff
@@ -0,0 +1,295 @@
+---
+layout: post
+title: Mini app open source projects
+authors:
+  - thomassteiner
+date:  2021-02-05
+# updated: 2021-01-25
+description: |
+  This chapter presents a selection of interesting mini app open source projects.
+tags:
+  - mini-apps
+---
+
+## kbone
+
+The [kbone](https://wechat-miniprogram.github.io/kbone/docs/) project
+([open source on GitHub](https://github.com/Tencent/kbone)) implements an adapter that simulates a
+browser environment in the adaptation layer, so that code written for the web can run without
+changes in a mini app. Several starter templates (among them
+[Vue](https://github.com/wechat-miniprogram/kbone-template-vue),
+[React](https://github.com/wechat-miniprogram/kbone-template-react), and
+[Preact](https://github.com/wechat-miniprogram/kbone-template-preact)) exist that make the
+onboarding experience for web developers coming from these frameworks easier.
+
+A new project can be created with the `kbone-cli` tool. A wizard asks what framework to initiate the
+project with. The code snippet below shows the Preact demo. In the code snippet below, the `mp`
+command builds the mini app, the `web` command builds the web app, and `build` creates the
+production web app.
+
+```bash
+npx kbone-cli init my-app
+cd my-app
+npm run mp
+npm run web
+npm run build
+```
+
+The code snippet below shows a simple counter component that then gets isomorphically rendered in a
+mini app and a web app. The overhead of the mini app is significant, purely judging from the DOM
+structure.
+
+```js
+import { h, Component } from "preact";
+import "./index.css";
+
+class Counter extends Component {
+  state = { count: 1 };
+
+  sub = () => {
+    this.setState((prevState) => {
+      return { count: --prevState.count };
+    });
+  };
+
+  add = () => {
+    this.setState((prevState) => {
+      return { count: ++prevState.count };
+    });
+  };
+
+  clickHandle = () => {
+    if ("undefined" != typeof wx && wx.getSystemInfoSync) {
+      wx.navigateTo({
+        url: "../log/index?id=1",
+      });
+    } else {
+      location.href = "log.html";
+    }
+  };
+
+  render({}, { count }) {
+    return (
+      <div>
+        <button onClick={this.sub}>-</button>
+        <span>{count}</span>
+        <button onClick={this.add}>+</button>
+        <div onClick={this.clickHandle}>跳转</div>
+      </div>
+    );
+  }
+}
+
+export default Counter;
+```
+
+<figure class="w-figure">
+  <img src="preact-wechat.png" alt="The Preact kbone template demo app opened in WeChat DevTools. Inspecting the DOM structure shows a significant overhead compared to the web app.">
+  <figcaption class="w-figure">
+   The Preact kbone template demo app opened in WeChat DevTools. Note the complex DOM structure and how the plus and minus buttons are actually not <code>&lt;button&gt;</code> elements.
```

Comment:
Makes me wonder about accessibility. Could be a good topic for a future update.

Reply:
Just like with Flutter :-)